packages/common/src/common.ts

```diff
@@ -285,57 +285,89 @@ export class Common extends EventEmitter {
     blockNumber = toType(blockNumber, TypeOutput.BigInt)
     td = toType(td, TypeOutput.BigInt)
 
-    let hardfork = Hardfork.Chainstart
-    let minTdHF
-    let maxTdHF
-    let previousHF
+    /** The hardfork config seen at hardfork before the merge */
+    let preMergeHF: HardforkConfig | undefined = undefined
+    /**
+     * Merge hardfork config if we have passed the merge hardfork while finding hardfork
+     * because next hardfork's block <= blockNumber
+     */
+    let passedMergeHF: HardforkConfig | undefined = undefined
+    /** Candiate hardfork config, updated as we iterate amongst the hardforks */
+    let hardfork: HardforkConfig | undefined = undefined
+
+    // Hardforks specified are strickly in the order of application, so we can make some intelligent decisions
+    // based on that especially regarding the merge hardfork where block might not be specified
     for (const hf of this.hardforks()) {
-      // Skip comparison for not applied HFs
-      if (hf.block === null) {
-        if (td !== undefined && td !== null && hf.ttd !== undefined && hf.ttd !== null) {
-          if (td >= BigInt(hf.ttd)) {
-            return hf.name
-          }
-        }
+      // Check if hf can be applied, if not then break and the last hardfork set is the response
+      if (hf.block === null && (hf.ttd === undefined || hf.ttd === null)) {
+        // Not a hardfork which has been scheduled yet so just skip
         continue
       }
-      if (blockNumber >= BigInt(hf.block)) {
-        hardfork = hf.name as Hardfork
-      }
-      if (td && (typeof hf.ttd === 'string' || typeof hf.ttd === 'bigint')) {
-        if (td >= BigInt(hf.ttd)) {
-          minTdHF = hf.name
-        } else {
-          maxTdHF = previousHF
+
+      if (hf.ttd !== undefined && hf.ttd !== null) {
+        // Ok, this is the hardfork with ttd
+        // If preMergeHF assigned, we have already seen merge hardfork
+        if (preMergeHF !== undefined) {
+          throw new Error(
+            `Invalid hardfork config with repeat ttd hardfork=${hf.name} ttd=${hf.ttd}`
+          )
         }
+        preMergeHF = hardfork
       }
-      previousHF = hf.name
-    }
-    if (td) {
-      let msgAdd = `block number: ${blockNumber} (-> ${hardfork}), `
-      if (minTdHF !== undefined) {
-        if (!this.hardforkGteHardfork(hardfork, minTdHF)) {
-          const msg = 'HF determined by block number is lower than the minimum total difficulty HF'
-          msgAdd += `total difficulty: ${td} (-> ${minTdHF})`
-          throw new Error(`${msg}: ${msgAdd}`)
+
+      // If merge hardfork has block null set, we will make determination if this is merge hardfork or preMergeHF
+      //  - if no future hardforks qualify by block number
+      // So we assume it to qualify for now to move to check future post merge hardfork qualitfication
+      //
+      // For the hardforks with block specified we check for qualification
```

Comment:
```suggestion
      // If merge fork has fork block set, we've found the correct hardfork
      // for `blockNumber` so break here.  Otherwise, check any additional forks.
```